compiler/rustc_infer/src/infer/error_reporting/suggest.rs

```diff
@@ -351,6 +351,73 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         }
     }
 
+    pub(super) fn suggest_function_pointers(
+        &self,
+        exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
+        diag: &mut Diagnostic,
+    ) {
+        let ty::error::ExpectedFound { expected, found } = exp_found;
+        if expected.peel_refs().is_fn() && found.peel_refs().is_fn() {
+            let expected_inner = expected.peel_refs();
+            let found_inner = found.peel_refs();
+            diag.note(
+                "different `fn` items always have unique types, even if their signatures are \
+                     the same",
+            );
+            match (&expected_inner.kind(), &found_inner.kind()) {
+                (ty::FnDef(did1, substs1), ty::FnDef(did2, substs2)) => {
+                    let sig1 = self.tcx.bound_fn_sig(*did1).subst(self.tcx, substs1);
+                    let sig2 = self.tcx.bound_fn_sig(*did2).subst(self.tcx, substs2);
+                    if sig1 != sig2 {
+                        return;
+                    }
+                    diag.help(&format!(
+                        "if the expected type is due to type inference, \
+                                       consider declaring the type as `{}` and casting the expected `fn` to a function pointer: \
+                                       `({} as {})`",
+                                       sig1,
+                        self.tcx.def_path_str_with_substs(*did1, substs1),
+                        sig1
+                    ));
+                }
+                (ty::FnDef(did, substs), ty::FnPtr(sig)) => {
+                    diag.help(&format!(
+                        "change the expected type to be a function pointer reference `&{}`",
+                        sig
+                    ));
+                    diag.help(&format!(
+                    "if the expected type is due to type inference, cast the expected `fn` to a function \
+                     pointer: `&({} as {})`",
+                    self.tcx.def_path_str_with_substs(*did, substs), sig));
```

Comment:
`self.tcx.def_path...` is equal to just printing out the `expected` ty you're matching on. Same for all the other call-sites.

Reply:
printing `expected` prints `fn(u32) -> u32 {name_of_function}`, whereas the `self.tcx.def_path...` prints  `name_of_function`.